Add openstack provider

just1not2 opened the pull request:
# Short description

This PR aims at adding the Openstack provider and the connection to an Openstack cluster


# Changes

* add provider base support
* add the openstack provider (with crud)
* add pylint conf


# Tests

List, create and delete instances on the Openstack cluster

just1not2 committed:
🎨(providers) create provider directories

We need to separate clearly each providers to facilitate
the contributions with other providers.

just1not2 committed:
🔨(providers) add provider env parameter

We need to be able to switch between providers easily, and
to install the right dependencies when needed.

just1not2 committed:
🔨(lint) add pylint support

We should use a linter to make sure our code keeps
being clean!

just1not2 committed:
🐛(requirements) fix pip3 path

Pip packages were not installed properly.

just1not2 committed:
✨(providers) implement dynamic provider import

We must load the provider that is specific to the
configuration.

just1not2 committed:
🎨(provider) add provider service schema

We must make the provider service classes unified
so we made a base class.

just1not2 committed:
✨(openstack) add base openstack provider

We need to setup the connection to the openstack
cluster.

just1not2 committed:
🔧(env) move openstack conf to .env

We should set all parameters to the environment
file to facilitate configuration.

just1not2 committed:
🔧(env) add metadata parameters

We need to distinct VMs that are gateways
and other VMs.

just1not2 committed:
✨(openstack) add creation method

We set new openstack variables to describe
used resources.

just1not2 committed:
✨(openstack) implement delete functionality

The schema was also modified.

just1not2 committed:
🔧(env) update default env value metadata

We should not mention gateways on this project
which is totally independant of any support.

just1not2 committed:
📝(env) add new env vars to README

All Openstack variables should be added to the
README for documentation.

just1not2 committed:
✏️(global fix missing new line

To make it more readable.

Benjou02 commented:
Not really related to the content of this PR, but it would be nice to have the __pycach__ directories in the .gitignore

Benjou02 commented on src/providers/openstack/main.py:
```diff
@@ -0,0 +1,63 @@
+                OPENSTACK_METADATA_KEY in server["metadata"] and
+                server["metadata"][OPENSTACK_METADATA_KEY] == OPENSTACK_METADATA_VALUE
+            ):
+                servers[server["id"]] = server["addresses"][OPENSTACK_NETWORK][0]["addr"]
```
This line may depend on the order of the IPv6 and IPv4 addresses: to be safe, wouldn't it be better to match a specific one ? Like always take `"version": 4` for example

just1not2 committed:
🙈(gitignore) ignore python cache

In case your python is configured to spawn
pycache folders, they won't be commited.

just1not2 committed:
🔧(env) add ip version var

We might need a specific IP version to contact
the gateways.
